Support 'non-contiguous with holes' NJTs for contiguous clone()

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131945
* __->__ #132776
* #131937
* #131704
* #131898

It's possible to construct an NJT with "holes" by specifying both `offsets` and `lengths` metadata. When `nt.clone(memory_format=torch.contiguous_format)` is called on such an NJT, the result should be an NJT without holes.

This PR fixes this in simplistic way using `unbind()`, which isn't really supported in `torch.compile`. The longer term solution involves writing a proper kernel to support this.

NB: Another limitation is that the returned NJT does not have the same ragged structure as the input. While we could manually hack the nested int registry (or update the union find when that lands), this is the first instance where a NJT with holes and an NJT without holes could have the same ragged structure, and getting those to play nicely together requires some fairly involved updates. For now, this PR punts on these updates until we can clean this up.

jbschlosser committed:
Support 'non-contiguous with holes' NJTs for contiguous clone()

[ghstack-poisoned]

soulitzer commented on torch/testing/_internal/opinfo/definitions/nested.py:
```diff
@@ -205,6 +205,27 @@ def _f(op, sample):
+            values, offsets=offsets, lengths=lengths
+        )
+
+        yield SampleInput(njt, kwargs={"memory_format": memory_format})
```
Just to double check, would this test fail even if we didn't properly handle the contiguous_format via unbinding?
Maybe I'm missing something, but it seems like when we call `assertEqualIgnoringNestedInts`, it seems like we would end up unbinding anyway.

jbschlosser replied:
yep, I verified that the test fails before my `contiguous_format` handling logic. The reason why is that before this PR, the lengths metadata wouldn't be propagated and so unbinding would just give you the wrong results (not taking into account lengths).

soulitzer replied:
I see, thanks

ani300 reviewed (approved):
lgtm

ani300 commented on torch/nested/_internal/ops.py:
```diff
@@ -421,6 +421,37 @@ def is_contiguous_general(func, *args, **kwargs):
+            assert (
+                inp._ragged_idx == 1
+            ), "NJT with ragged_idx != 1 not supported for contiguous clone"
+            contig, _ = jagged_from_list(inp.unbind(), offsets=None)
```
this would be as "simple" as having the dim that used to be referenced by the lengths be used by the offsets here, right?

ani300 replied:
(referring to the TODO)

jbschlosser replied:
almost (I tried this briefly), I think there's more work required throughout to actually support ops with inputs that have the same nested int but different contiguity. We have to be smarter about how we operate on the `values` buffer when the lengths metadata is present.

ani300 replied:
I see... I guess it can wait for a further PR then

soulitzer reviewed (approved):
Sounds good!

jbschlosser committed:
Update on "Support 'non-contiguous with holes' NJTs for contiguous clone()"


It's possible to construct an NJT with "holes" by specifying both `offsets` and `lengths` metadata. When `nt.clone(memory_format=torch.contiguous_format)` is called on such an NJT, the result should be an NJT without holes.

This PR fixes this in simplistic way using `unbind()`, which isn't really supported in `torch.compile`. The longer term solution involves writing a proper kernel to support this.

NB: Another limitation is that the returned NJT does not have the same ragged structure as the input. While we could manually hack the nested int registry (or update the union find when that lands), this is the first instance where a NJT with holes and an NJT without holes could have the same ragged structure, and getting those to play nicely together requires some fairly involved updates. For now, this PR punts on these updates until we can clean this up.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge
